lib/charms/tls_certificates_interface/v2/tls_certificates.py

```diff
@@ -1159,6 +1159,57 @@ def _revoke_certificates_for_which_no_csr_exists(self, relation_id: int) -> None
                 )
                 self.remove_certificate(certificate=certificate["certificate"])
 
+    def get_requirer_csrs_with_no_certs(self) -> List[str]:
+        """Filters requirer CSRs for which no certificate exists.
+
+        Returns:
+            list: List of requirer CSRs for which no certificate exists.
+        """
+        all_csrs = copy.deepcopy(self.get_requirer_csrs)
+        for unit_csrs in all_csrs:
+            for csr in unit_csrs["unit_csrs"]:
+                if (
+                    csr
+                    in self.get_issued_certificates(csr["relation_id"])[
+                        csr["application_name"]
+                    ].keys()
+                ):
+                    unit_csrs["unit_csrs"].remove(csr)
+            if len(unit_csrs["unit_csrs"]) == 0:
+                all_csrs.remove(unit_csrs)
+        return all_csrs
+
+    def get_requirer_csrs(
+        self, relation_id: Optional[int] = None
+    ) -> List[Dict[str, Union[str, List[str]]]]:
+        """Returns a list of requirers' CSRs.
+
+        It returns CSRs from all relations if relation_id is not specified.
+        CSRs are returned per relation id, application name and unit name.
+
+        Returns:
+            list: CSRs per relation id, application name and unit name.
+        """
+        csrs: List[Dict[str, Union[str, List[str]]]] = []
+        relations = (
+            [self.model.relations[self.relationship_name][relation_id]]
+            if relation_id
+            else self.model.relations.get(self.relationship_name, [])
+        )
+        for relation in relations:
+            for unit in relation.units:
+                requirer_relation_data = _load_relation_data(relation.data[unit])
+                unit_csrs_list = requirer_relation_data.get("certificate_signing_requests", [])
+                csrs.append(
+                    {
+                        "relation_id": relation.id,
+                        "application_name": relation.app.name,
+                        "unit_name": unit.name,
+                        "unit_csrs": unit_csrs_list,
+                    }
+                )
+        return csrs
```

Comment:
In this method what we call a CSR is a dict containing relation id, application name, etc. In the other method we're only returning the string for the CSR. We should standardize around what is returned.